driver-core/src/main/java/com/datastax/driver/core/Cluster.java

```diff
@@ -2282,6 +2288,230 @@ public void run() {
+
+        }
+
+        private class NodeRefreshRequestDeliveryCallback implements EventDebouncer.DeliveryCallback<NodeRefreshRequest> {
+
+            @Override
+            public ListenableFuture<?> deliver(List<NodeRefreshRequest> events) {
+                Map<Host, HostEvent> hosts = new HashMap<Host, HostEvent>();
+                // only keep the last event for each host
+                for (NodeRefreshRequest req : events) {
+                    hosts.put(req.host, req.eventType);
+                }
+                List<ListenableFuture<?>> futures = new ArrayList<ListenableFuture<?>>(hosts.size());
+                for (final Entry<Host, HostEvent> entry : hosts.entrySet()) {
+                    final Host host = entry.getKey();
+                    HostEvent eventType = entry.getValue();
+                    switch (eventType) {
+                        case UP:
+                            schedule(hostUp(host));
+                            break;
+                        case ADDED:
+                            schedule(hostAdded(host));
+                            break;
+                        case DOWN:
+                            execute(hostDown(host));
+                            break;
+                        case REMOVED:
+                            execute(hostRemoved(host));
+                            break;
+                    }
+                }
+                return Futures.allAsList(futures);
+            }
+
+            private ListenableFuture<?> execute(ExceptionCatchingRunnable task) {
+                return executor.submit(task);
+            }
+
+            private ListenableFuture<?> schedule(final ExceptionCatchingRunnable task) {
+                // Cassandra tends to send notifications for new/up nodes a bit early (it is triggered once
+                // gossip is up, but that is before the client-side server is up), so we add a delay
+                // (otherwise the connection will likely fail and have to be retry which is wasteful). This
+                // probably should be fixed C* side, after which we'll be able to remove this.
+                final SettableFuture<?> future = SettableFuture.create();
+                scheduledTasksExecutor.schedule(new ExceptionCatchingRunnable() {
+                    public void runMayThrow() throws Exception {
+                        ListenableFuture<?> f = execute(task);
+                        Futures.addCallback(f, new FutureCallback<Object>() {
+                            @Override
+                            public void onSuccess(Object result) {
+                                future.set(null);
+                            }
+
+                            @Override
+                            public void onFailure(Throwable t) {
+                                future.setException(t);
+                            }
+                        });
+                    }
+                }, NEW_NODE_DELAY_SECONDS, TimeUnit.SECONDS);
```

Comment:
This could be somewhat problematic for our tests since we set a really large NEW_NODE_DELAY_SECONDS, though maybe we should stop doing that.